src/librustc_middle/ty/query/plumbing.rs

```diff
@@ -380,8 +380,8 @@ macro_rules! define_queries_inner {
         impl TyCtxtEnsure<$tcx> {
             $($(#[$attr])*
             #[inline(always)]
-            pub fn $name(self, key: $K) {
-                ensure_query::<queries::$name<'_>, _>(self.tcx, key)
+            pub fn $name(self, key: impl Into<$K>) {
+                ensure_query::<queries::$name<'_>, _>(self.tcx, key.into())
```

Comment:
I'd make some new methods that call the existing ones in this file to avoid making these generic. We want to make sure that `ensure_query` and `get_query` is only instantiated in `rustc_middle.`

Reply:
Just that i understand; the proposal here is to add new methods _only_ for methods where the key is a `DefId` and give them a name similar to the original one but with a prefix (for example). We would end up with methods like `local_generics_of` for `LocalDefId` instead of `generics_of` for `DefId`?

If so, all callers that have a `LocalDefId` would have to use the new methods, and then there is not much benefit compared to calling `.to_def_id()` everywhere. All code will have to be changed anyway, either to call the new methods or to call `.to_def_id()`.

Reply:
I believe they want the caller of `ensure_query` to remain monomorphic and for you to introduce a generic wrapper around it that calls `.into()` on the query param before passing it to the monomorphic version. This requires creating a new `$name` somewhere.